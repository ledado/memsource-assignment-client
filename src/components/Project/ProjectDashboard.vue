<template>
  <v-row no-gutters>
    <v-col cols="12" class="text-h4">
      {{ $t('project.overview.title') }}
    </v-col>
    <v-col cols="12">
      <v-row>
        <v-col cols="4">
          <project-statistic-card
            :value="countOfNewProjects"
            :title="$t('project.status.new')"
            :description="$t('project.overview.new.description')"
            background-class="blue"
          ></project-statistic-card>
        </v-col>
        <v-col cols="4">
          <project-statistic-card
            :value="countOfCompletedProjects"
            :title="$t('project.status.completed')"
            :description="$t('project.overview.completed.description')"
            background-class="orange"
          ></project-statistic-card>
        </v-col>
        <v-col cols="4">
          <project-statistic-card
            :value="countOfDeliveredProjects"
            :title="$t('project.status.delivered')"
            :description="$t('project.overview.delivered.description')"
            background-class="green"
          ></project-statistic-card>
        </v-col>
      </v-row>
    </v-col>
    <v-col cols="12">
      <v-row justify="space-between">
        <v-col cols="auto" class="text-h4">
          {{ $t('project.list.title') }}
        </v-col>
        <v-col cols="auto">
          <v-btn color="success" :to="{name: $data.routerTypes.PROJECT_NEW}">
            {{ $t('project.list.create_new') }}
          </v-btn>
        </v-col>
      </v-row>
    </v-col>
    <v-col cols="12">
      <v-row>
        <v-col cols="auto">
          <v-select
            :items="nameFilterItems"
            v-model="selectedNameFilter"
            :label="$t('project.list.filter_by_name')"
            hide-details
            outlined
            multiple
          ></v-select>
        </v-col>
        <v-col cols="auto">
          <v-select
            :items="statusFilterItems"
            v-model="selectedStatusFilter"
            :label="$t('project.list.filter_by_status')"
            hide-details
            outlined
            multiple
          ></v-select>
        </v-col>
        <v-col cols="auto">
          <v-select
            :items="sortItems"
            v-model="projectsSortBy"
            :label="$t('project.list.sort_by.label')"
            hide-details
            outlined
          >
          </v-select>
        </v-col>
        <v-col cols="auto">
          <v-checkbox v-model="projectsSortDesc" :label="$t('project.list.sort_by_desc')"></v-checkbox>
        </v-col>
      </v-row>
    </v-col>
    <v-col cols="12" class="mt-3">
      <v-data-table
        :headers="projectHeaders"
        :items="filteredProjects"
        :items-per-page="5"
        :sort-by.sync="projectsSortBy"
        :sort-desc.sync="projectsSortDesc"
        :loading="loading"
      >
        <template v-slot:item.dateCreated="{item}">
          {{ item.dateCreated | formatDate('HH:mm, DD.MM.YYYY') }}
        </template>
        <template v-slot:item.dateUpdated="{item}">
          {{ item.dateUpdated | formatDate('HH:mm, DD.MM.YYYY') }}
        </template>
        <template v-slot:item.dateDue="{item}">
          {{ item.dateDue | formatDate('HH:mm, DD.MM.YYYY') }}
        </template>
        <template v-slot:item.actions="{item}">
          <v-icon small class="mr-2" @click="editProject(item)">
            mdi-pencil
          </v-icon>
        </template>
      </v-data-table>
    </v-col>
  </v-row>
</template>

<script lang="ts">
import {Component, Vue} from 'vue-property-decorator';
import ProjectService from '@/service/api/ProjectService';
import {ProjectItem, ProjectStatusEnum} from '@/service/api/types/ProjectTypes';
import ProjectStatisticCard from '@/components/Project/ProjectStatisticCard.vue';

@Component({
  components: {ProjectStatisticCard},
})
export default class ProjectDashboard extends Vue {
  projects: ProjectItem[] = [];
  loading: boolean = false;

  projectHeaders: any = [];

  projectsSortBy = 'id';
  projectsSortDesc = false;

  sortItems: any = [];

  nameFilterItems: string[] = [];
  selectedNameFilter: string[] = [];

  statusFilterItems = [ProjectStatusEnum.NEW, ProjectStatusEnum.COMPLETED, ProjectStatusEnum.DELIVERED];
  selectedStatusFilter: ProjectStatusEnum[] = [];

  created() {
    this.sortItems = [
      {value: 'id', text: this.$t('project.list.sort_by.id').toString()},
      {value: 'name', text: this.$t('project.list.sort_by.name').toString()},
      {value: 'status', text: this.$t('project.list.sort_by.status').toString()},
      {value: 'dateDue', text: this.$t('project.list.sort_by.dateDue').toString()},
    ];
    this.projectHeaders = [
      {
        text: 'Id',
        align: 'start',
        value: 'id',
        sortable: false,
      },
      {text: this.$t('project.list.table.name').toString(), value: 'name', sortable: false},
      {text: this.$t('project.list.table.source_language').toString(), value: 'sourceLanguage', sortable: false},
      {text: this.$t('project.list.table.status').toString(), value: 'status', sortable: false},
      {text: this.$t('project.list.table.target_languages').toString(), value: 'targetLanguages', sortable: false},
      {text: this.$t('project.list.table.date_created').toString(), value: 'dateCreated', sortable: false},
      {text: this.$t('project.list.table.date_updated').toString(), value: 'dateUpdated', sortable: false},
      {text: this.$t('project.list.table.date_due').toString(), value: 'dateDue', sortable: false},
      {text: this.$t('project.list.table.actions').toString(), value: 'actions', sortable: false},
    ];
    this.loading = true;
    ProjectService.getAll()
      .then((response) => {
        this.projects = response.data._embedded.projects;
        this.nameFilterItems = this.projects.map((project) => project.name);
      })
      .catch(() => {
        this.$toasted.error(this.$t('global.error.request').toString());
      })
      .finally(() => {
        this.loading = false;
      });
  }

  get countOfNewProjects() {
    return this.projects.filter((project) => project.status === ProjectStatusEnum.NEW).length;
  }

  get countOfCompletedProjects() {
    return this.projects.filter((project) => project.status === ProjectStatusEnum.COMPLETED).length;
  }

  get countOfDeliveredProjects() {
    return this.projects.filter((project) => project.status === ProjectStatusEnum.DELIVERED).length;
  }

  get filteredProjects() {
    let filteredItems = this.projects;
    if (this.selectedStatusFilter.length > 0) {
      filteredItems = filteredItems.filter((item) => this.selectedStatusFilter.includes(item.status));
    }
    if (this.selectedNameFilter.length > 0) {
      filteredItems = filteredItems.filter((item) => this.selectedNameFilter.includes(item.name));
    }
    return filteredItems;
  }

  editProject(project: ProjectItem) {
    this.$router.push({name: this.$data.routerTypes.PROJECT_UPDATE, params: {id: project.id.toString()}});
  }
}
</script>
<style lang="scss" scoped></style>
