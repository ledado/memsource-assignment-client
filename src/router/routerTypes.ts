export const routerTypes = {
  MAIN: 'main',
  PROJECT_NEW: 'project-new',
  PROJECT_UPDATE: 'project-update',
};
