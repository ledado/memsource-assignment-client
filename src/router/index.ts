import Vue from 'vue';
import VueRouter, {RouteConfig} from 'vue-router';
import {errorBus} from '@/service/events/ErrorBus';
import {layoutTypes} from '@/layouts/layoutTypes';
import {routerTypes} from '@/router/routerTypes';

Vue.use(VueRouter);

const routes: Array<RouteConfig> = [
  {
    path: '/',
    name: routerTypes.MAIN,
    component: () => import('../views/Main.vue'),
    meta: {
      layout: layoutTypes.DEFAULT,
    },
  },
  {
    path: '/project/new',
    name: routerTypes.PROJECT_NEW,
    component: () => import('../views/Project/MainProjectCreate.vue'),
    meta: {
      layout: layoutTypes.DEFAULT,
    },
  },
  {
    path: '/project/:id',
    name: routerTypes.PROJECT_UPDATE,
    component: () => import('../views/Project/MainProjectUpdate.vue'),
    meta: {
      layout: layoutTypes.DEFAULT,
    },
    props: true,
  },
  {
    path: '*',
    component: () => import('../views/Error/Error404.vue'),
  },
];

const router = new VueRouter({
  mode: 'history',
  base: process.env.BASE_URL,
  routes,
});

// error state clean up
router.afterEach((to, from) => {
  if (from.name && errorBus.state) {
    errorBus.errorCode = undefined;
    errorBus.state = false;
  }
});

export default router;
