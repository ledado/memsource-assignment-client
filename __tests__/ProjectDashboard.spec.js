import {shallowMount} from '@vue/test-utils';
import ProjectDashboard from '@/components/Project/ProjectDashboard.vue';
import {routerTypes} from '@/router/routerTypes';

describe('ProjectDashboard.vue', () => {
  it('render component', () => {
    const wrapper = shallowMount(ProjectDashboard, {
      stubs: ['router-link'],
      mocks: {
        $t: (msg) => msg,
      },
      mixins: [
        {
          data: () => {
            return {
              routerTypes,
            };
          },
        },
      ],
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
