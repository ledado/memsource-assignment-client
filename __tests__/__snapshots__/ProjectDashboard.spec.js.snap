// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectDashboard.vue render component 1`] = `
"<v-row-stub tag=\\"div\\" nogutters=\\"true\\">
  <v-col-stub cols=\\"12\\" tag=\\"div\\" class=\\"text-h4\\">
    project.overview.title
  </v-col-stub>
  <v-col-stub cols=\\"12\\" tag=\\"div\\">
    <v-row-stub tag=\\"div\\">
      <v-col-stub cols=\\"4\\" tag=\\"div\\">
        <project-statistic-card-stub value=\\"0\\" title=\\"project.status.new\\" description=\\"project.overview.new.description\\" backgroundclass=\\"blue\\"></project-statistic-card-stub>
      </v-col-stub>
      <v-col-stub cols=\\"4\\" tag=\\"div\\">
        <project-statistic-card-stub value=\\"0\\" title=\\"project.status.completed\\" description=\\"project.overview.completed.description\\" backgroundclass=\\"orange\\"></project-statistic-card-stub>
      </v-col-stub>
      <v-col-stub cols=\\"4\\" tag=\\"div\\">
        <project-statistic-card-stub value=\\"0\\" title=\\"project.status.delivered\\" description=\\"project.overview.delivered.description\\" backgroundclass=\\"green\\"></project-statistic-card-stub>
      </v-col-stub>
    </v-row-stub>
  </v-col-stub>
  <v-col-stub cols=\\"12\\" tag=\\"div\\">
    <v-row-stub tag=\\"div\\" justify=\\"space-between\\">
      <v-col-stub cols=\\"auto\\" tag=\\"div\\" class=\\"text-h4\\">
        project.list.title
      </v-col-stub>
      <v-col-stub cols=\\"auto\\" tag=\\"div\\">
        <v-btn-stub color=\\"success\\" tag=\\"button\\" activeclass=\\"\\" to=\\"[object Object]\\" type=\\"button\\">
          project.list.create_new
        </v-btn-stub>
      </v-col-stub>
    </v-row-stub>
  </v-col-stub>
  <v-col-stub cols=\\"12\\" tag=\\"div\\">
    <v-row-stub tag=\\"div\\">
      <v-col-stub cols=\\"auto\\" tag=\\"div\\">
        <v-select-stub errorcount=\\"1\\" errormessages=\\"\\" messages=\\"\\" rules=\\"\\" successmessages=\\"\\" value=\\"\\" appendicon=\\"$dropdown\\" backgroundcolor=\\"\\" hidedetails=\\"true\\" label=\\"project.list.filter_by_name\\" loaderheight=\\"2\\" clearicon=\\"$clear\\" outlined=\\"true\\" type=\\"text\\" valuecomparator=\\"[Function]\\" nodatatext=\\"$vuetify.noDataText\\" items=\\"\\" itemcolor=\\"primary\\" itemdisabled=\\"disabled\\" itemtext=\\"text\\" itemvalue=\\"value\\" menuprops=\\"[object Object]\\" multiple=\\"true\\"></v-select-stub>
      </v-col-stub>
      <v-col-stub cols=\\"auto\\" tag=\\"div\\">
        <v-select-stub errorcount=\\"1\\" errormessages=\\"\\" messages=\\"\\" rules=\\"\\" successmessages=\\"\\" value=\\"\\" appendicon=\\"$dropdown\\" backgroundcolor=\\"\\" hidedetails=\\"true\\" label=\\"project.list.filter_by_status\\" loaderheight=\\"2\\" clearicon=\\"$clear\\" outlined=\\"true\\" type=\\"text\\" valuecomparator=\\"[Function]\\" nodatatext=\\"$vuetify.noDataText\\" items=\\"NEW,COMPLETED,DELIVERED\\" itemcolor=\\"primary\\" itemdisabled=\\"disabled\\" itemtext=\\"text\\" itemvalue=\\"value\\" menuprops=\\"[object Object]\\" multiple=\\"true\\"></v-select-stub>
      </v-col-stub>
      <v-col-stub cols=\\"auto\\" tag=\\"div\\">
        <v-select-stub errorcount=\\"1\\" errormessages=\\"\\" messages=\\"\\" rules=\\"\\" successmessages=\\"\\" value=\\"id\\" appendicon=\\"$dropdown\\" backgroundcolor=\\"\\" hidedetails=\\"true\\" label=\\"project.list.sort_by.label\\" loaderheight=\\"2\\" clearicon=\\"$clear\\" outlined=\\"true\\" type=\\"text\\" valuecomparator=\\"[Function]\\" nodatatext=\\"$vuetify.noDataText\\" items=\\"[object Object],[object Object],[object Object],[object Object]\\" itemcolor=\\"primary\\" itemdisabled=\\"disabled\\" itemtext=\\"text\\" itemvalue=\\"value\\" menuprops=\\"[object Object]\\"></v-select-stub>
      </v-col-stub>
      <v-col-stub cols=\\"auto\\" tag=\\"div\\">
        <v-checkbox-stub errorcount=\\"1\\" errormessages=\\"\\" messages=\\"\\" rules=\\"\\" successmessages=\\"\\" backgroundcolor=\\"\\" label=\\"project.list.sort_by_desc\\" ripple=\\"true\\" valuecomparator=\\"[Function]\\" indeterminateicon=\\"$checkboxIndeterminate\\" officon=\\"$checkboxOff\\" onicon=\\"$checkboxOn\\"></v-checkbox-stub>
      </v-col-stub>
    </v-row-stub>
  </v-col-stub>
  <v-col-stub cols=\\"12\\" tag=\\"div\\" class=\\"mt-3\\">
    <v-data-table-stub mobilebreakpoint=\\"600\\" items=\\"\\" options=\\"[object Object]\\" sortby=\\"id\\" customsort=\\"[Function]\\" page=\\"1\\" itemsperpage=\\"5\\" groupby=\\"\\" groupdesc=\\"\\" customgroup=\\"[Function]\\" locale=\\"en-US\\" customfilter=\\"[Function]\\" serveritemslength=\\"-1\\" itemkey=\\"id\\" value=\\"\\" expanded=\\"\\" loading=\\"true\\" noresultstext=\\"$vuetify.dataIterator.noResultsText\\" nodatatext=\\"$vuetify.noDataText\\" loadingtext=\\"$vuetify.dataIterator.loadingText\\" selectablekey=\\"isSelectable\\" loaderheight=\\"4\\" headers=\\"[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]\\" expandicon=\\"$expand\\" itemclass=\\"\\"></v-data-table-stub>
  </v-col-stub>
</v-row-stub>"
`;
